### PR TITLE
Create github workflow for publish container

### PR DESCRIPTION
**Pull request by andreeaflorescu**
### Summary of the PR

Make it possible to automatically publish docker containers.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: build_container.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
```
Comment (Ablu) [inline build_container.sh:1]: you might want to add a `set -e` here in order to abort if things fail (`-x` might also make sense for better logs)

File: Dockerfile [hunk 2]
```diff
@@ -1,58 +1,8 @@
...
-
-# Install libgpiod (required by vhost-device crate)
-RUN cd /opt/ && \
-    git clone --depth 1 --branch v2.0 https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.git/ && \
-    cd libgpiod && ./autogen.sh --prefix=/usr && make && make install; \
-    cd
-
-RUN echo "{\"rev\":\"$GIT_COMMIT\",\"branch\":\"${GIT_BRANCH}\"}" > /buildinfo.json
+COPY build_container.sh /opt/src/scripts/build_container.sh
+RUN chmod +x /opt/src/scripts/build_container.sh && /opt/src/scripts/build_container.sh
```
Comment (Ablu) [inline Dockerfile:8]: The +x could probably also be applied to the file in the repo?

File: build_container.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+DEBIAN_FRONTEND="noninteractive" apt-get install --no-install-recommends -y \
+    curl gcc git python3 python3-pip shellcheck \
+    libssl-dev tzdata cmake g++ pkg-config jq libcurl4-openssl-dev libelf-dev \
+    libdw-dev binutils-dev libiberty-dev make \
+    cpio bc flex bison wget xz-utils fakeroot \
+    autoconf autoconf-archive automake libtool \
+    iproute2
+
+# cleanup
+apt-get clean && rm -rf /var/lib/apt/lists/* \
```
Comment (stefano-garzarella) [inline build_container.sh:15]: The last `\` can be removed

File: Dockerfile [hunk 4]
```diff
@@ -27,7 +27,7 @@ RUN apt-get update && \
     && rm -rf /var/lib/apt/lists/*
 
 # Install pytest and boto3.
-RUN pip3 install pytest pexpect boto3 pytest-timeout
+RUN pip3 install --no-cache-dir pytest pexpect boto3 pytest-timeout
```
[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply Dockerfile]: Yes, it is, good catch! I will mention it as well.

File: build_container.sh [hunk 5]
```diff
@@ -0,0 +1,42 @@
+set -ex
```
[the remark this answers is not in the document]
Comment (andreeaflorescu) [reply build_container.sh]: I think it should work now. I've updated it.

**Review by stefano-garzarella — APPROVED** [review]

**Review by lauralt — APPROVED** [review]

File: .github/workflows/docker-publish.yml [hunk 6]
```diff
@@ -0,0 +1,87 @@
+name: Docker
+
+on:
+  push:
+    branches: [ "main" ]
+    paths: Dockerfile
+  pull_request:
+    branches: [ "main" ]
```
Comment (rbradford) [inline .github/workflows/docker-publish.yml:8]: Not sure about this? I can't find documentation that indicates what this should do? PRs targetting that branch or?

**Review by rbradford — APPROVED** [review]

File: Dockerfile [hunk 7]
```diff
@@ -4,51 +4,5 @@ ARG RUST_TOOLCHAIN="1.67.1"
...
-    curl gcc git python3 python3-pip shellcheck \
-    # kcov dependencies
-    libssl-dev tzdata cmake g++ pkg-config jq libcurl4-openssl-dev libelf-dev \
-    libdw-dev binutils-dev libiberty-dev make \
-    # utilities to build kernels
-    cpio bc flex bison wget xz-utils fakeroot \
-    # utilities to build usersapce packages
-    autoconf autoconf-archive automake libtool \
-    # bindgen dependency
-    libclang-dev \
```
Comment (vireshk) [inline Dockerfile:23]: @andreeaflorescu This is required to run bindgen for libgpiod crate, which is done while buidling vhost-device crate. This commit broke vhost-device:

https://github.com/rust-vmm/vhost-device/pull/308

**Pull request merged**